### DISCUSSION TITLE
Airmega 250 and 250S

### DISCUSSION THREAD

**Discussion by catduckgnaf** [RobertD502/home-assistant-iocare#28] [Q&A] [answered]
I know they are not on the "confirmed" working list.
Is there any reason they shouldn't work? Looking at buying one.

**Comment by RobertD502** [comment]
I don't know what the 250S purifiers return by the API, but there is a good chance they work. I can't say for certain as no one has come to me and told me they are using the 250S with this integration. Also, note the 250 won't work as it doesn't have any WiFi capabilities.

**Reply by catduckgnaf** [reply]
Thanks for moving to a discussion for me. Well I just ordered one as the wife likes how the 250S looks more then 300S and 400S.

When I get the 250S I will report back the working status, and if it doesn't I'll let you know and work to debug and figure it out.

Thanks for your project!

**Reply by RobertD502** [reply]
It should work given that even the non 300/400 models all have a low, medium, high speed and the same auto/night mode.

Only thing that may be different is the particle sensing and AQI. Some of the newer models don't return an AQI in the API response. If that is the case, you can always provide access to your purifier temporarily and I can check to see what is returned/what needs to be modified in the Home Assistant code and/or the backend library to get things perfectly aligned with your model.

**Comment by pcmoore** [comment, marked as answer]
I've had a 250S for several months now and can report that it works with using this integration.

Thanks for all your work on this @RobertD502!

**Reply by RobertD502** [reply]
Thanks for letting me know. I'll get it added to the list.

**Reply by catduckgnaf** [reply]
Excellent! Glad I asked. Does the AQI work?

**Reply by pcmoore** [reply]
Here is a quick screenshot of the device view of the 250S, minus the automations, scenes, and scripts panels; it basically looks the same as my 400S units and behaves the same as well.  If you have any additional questions let me know.

![screenshot_redacted](https://github.com/RobertD502/home-assistant-iocare/assets/8494929/d772177a-d1b1-426c-8ff7-447245ae5e5f)

**Reply by RobertD502** [reply]
Would you mind turning on debug logging? At the moment I have turned off adding the AQI sensor for purifiers that have a MCU version (these weren't returning a valid AQI for the 300S and 400S). I'd be curious to know if your 250S returns an AQI even with a MCU version present.

The debug log should contain the data returned by your purifiers.

**Comment by pcmoore** [comment]
> Would you mind turning on debug logging?

Sure, but I would appreciate a pointer on how to do that and what I should send back :)

> At the moment I have turned off adding the AQI sensor for purifiers that have a MCU version (these weren't returning a valid AQI for the 300S and 400S).

Ah ha!  This answers a minor question I've had for some time.  In addition to the 250S, we also have two 400S units and one reports an AQI number in the integration while the other does not; similarly, in the official app one reports version numbers while the other does not.  I wasn't too bothered by this as the purifiers work normally in every other way, but I was curious if something was amiss, but it looks like I'm not alone.  There must have been a small design change, or firmware revision, part way through production on these units.

**Reply by RobertD502** [reply]
From your computer, on the Home Assistant integrations page, click on the Coway integration. Under the Coway logo, you'll see the option `Enable debug logging`. Click on it. In order for debugging to start, you'll need to restart Home Assistant. Once you are back up and running, give it a few seconds and then navigate back to the Integrations page, click on Coway, and then disable debug logging. At this point a log file should download. Share that log here with me.


Regarding AQI: Although the "newer" 300S and 400S models are essentially the same, for some reason the ones with a MCU version returned by the API, don't return an AQI - the key is returned by the API, but there is no actual value. The common denominator was that devices, as it applies to the 300S and 400S, with a MCU version are the ones not returning an AQI, which is why that sensor isn't created for those devices. This log will allow me to see if by any chance the 250S does in fact return an AQI even with a MCU version defined.

**Reply by pcmoore** [reply]
See attached.

[home-assistant_coway_2023-10-10T21-39-27.604Z-redacted.log](https://github.com/RobertD502/home-assistant-iocare/files/12862020/home-assistant_coway_2023-10-10T21-39-27.604Z-redacted.log)

**Reply by RobertD502** [reply]
As I suspected, the 250S returns an empty air quality index.

**Comment by mcleanrs** [comment]
I just set this up for my 250S. The basics are working great so far. Thank you / hooray!

One question -- are local changes reflected back into Home Assistant? In my brief tinkering it didn't seem like they're showing, but I might just be impatient or might need a power cycle of the purifier and/or Home Assistant. Wanted to confirm what the expected behavior is before I tinker too much. Thanks!

**Reply by RobertD502** [reply]
Local changes should be reported back to the Coway servers. When the integration polls the API and if the change you made is reflected on their servers, you will see the updated state in Home Assistant. You aren't the only one to notice this...sometimes changes made locally or even with the integration don't get reflected. This is a quirk in their system and usually turning the WiFi off and back on (on the purifier) forces the purifier to send its updated state to Coway's servers. Most of the time this happens if you are attempting to send too many commands back to back.

**Comment by jerrice** [comment]
I've been using this integration for the 250S for a while now, and it works great, thank you! A couple of notes though:
1) PM 2.5 does get reported back to the app, and used to work with this integration, but I think it needs to be enabled for this model as it no longer shows up at all in Home Assistant.
2) I think it is possible some of the confusion around state being reported back may be because it is somehow opposite of the setting? IOW, if I turn the "Light" button to off, it actually turns them on. If I turn it on, it actually turns it off. For my automations I just have to use the opposite of what is expected, but it is odd.

**Reply by RobertD502** [reply]
~~Turn on debug logging and post the results here. This will let me know if the air quality endpoint reports a 2.5 value.~~

Disregard: I see from a previous user's log that the 250s does report 2.5 particulates.

Regarding the light: this is probably a mixup between what the purifier has stored internally and the coway servers. Turn off the WiFi on your purifier and then turn it on again. This forces the purifier to resend its current state back to the coway servers.

**Reply by jerrice** [reply]
Great! I'll try cycling the wifi, and let you know.

**Reply by celeduc** [reply]
Did cycling the wifi work for you? It hasn't for me. I'm wondering whether I need to re-register the device with IOCare.

**Reply by jerrice** [reply]
No it did not for me. The IOCare app also shows the correct state all the time, leading me to believe that the API is just opposite for the 250S for some reason?

Also, I did download the latest version of the integration, but it looks like 10 was disabled? 250s seems to have both 2.5 and 10 reporting.

**Reply by RobertD502** [reply]
> No it did not for me. The IOCare app also shows the correct state all the time, leading me to believe that the API is just opposite for the 250S for some reason?
> 
> 
> 
> Also, I did download the latest version of the integration, but it looks like 10 was disabled? 250s seems to have both 2.5 and 10 reporting.

Yup, that's my bad on the PM 10 sensor! Will get it back in the next release.

Regarding the light, if you're willing to share your login credentials temporarily, I can investigate how the API handles setting the light for the 250S - send a DM on discord (my discord handle is the same as here - RobertD502)

**Reply by RobertD502** [reply]
@jerrice The issue with the light has been resolved in release `0.4.0`. The reason the light switch entity didn't work as expected is because different values are used to set the light mode when dealing with a 250S purifier + the 250S light has more than just On and Off. The light switch has been removed for 250S purifiers and replaced with a light select entity which allows you to set the light to either On, Off, or AQI Off.

**Reply by jerrice** [reply]
That is awesome, thank you so much!! I'll get things updated this weekend and let you know how it goes!

**Reply by jerrice** [reply]
@RobertDRob All looks good! Thank you for your work here!!
